File: Python/Problem_112.py
def is_bouncy(n: int) -> bool:
    inc = True
    dec = True
    last = n % 10
    n //= 10

    while n > 0:
        d = n % 10
        if d < last:
            inc = False
        if d > last:
            dec = False
        last = d
        n //= 10

        if not inc and not dec:
            return True
    return False  # monotone

def solve() -> int:
    bouncy = 0
    n = 1
    while True:
        n += 1
        if n >= 100 and is_bouncy(n):
            bouncy += 1
        if bouncy * 100 == n * 99:
            return n

if __name__ == "__main__":
    print(solve())  # 1587000

File: Java/Problem_112.java
public class Main {
    static boolean isBouncy(int n) {
        boolean inc = true, dec = true;
        int last = n % 10;
        n /= 10;

        while (n > 0) {
            int d = n % 10;
            if (d < last) inc = false;
            if (d > last) dec = false;
            last = d;
            n /= 10;

            if (!inc && !dec) return true;
        }
        return false;
    }

    public static void main(String[] args) {
        int bouncy = 0;
        int n = 1;

        while (true) {
            n++;
            if (n >= 100 && isBouncy(n)) bouncy++;
            if ((long)bouncy * 100 == (long)n * 99) {
                System.out.println(n); // 1587000
                return;
            }
        }
    }
}
